src/imitation/data/types.py

```diff
@@ -22,7 +29,267 @@
+
+    @classmethod
+    def concatenate(cls, dictobs_list: Iterable["DictObs"], axis=0) -> "DictObs":
+        """Returns a single dictobs concatenating the arrays by key."""
+        return cls(
+            {
+                k: np.concatenate(arr_list, axis=axis)
+                for k, arr_list in cls._unravel(dictobs_list).items()
+            },
+        )
+
+
+# DicObs utilities
+
+
+Observation = Union[np.ndarray, DictObs]
+ObsVar = TypeVar("ObsVar", np.ndarray, DictObs)
+
+
+def assert_not_dictobs(x: Observation) -> np.ndarray:
+    """Typeguard to assert `x` is an array, not a DictObs."""
+    if isinstance(x, DictObs):
+        assert False, "Dictionary observations are not supported here."
+    return x
+
+
+def concatenate_maybe_dictobs(arrs: List[ObsVar]) -> ObsVar:
+    """Concatenates a list of observations appropriately (depending on type)."""
+    assert len(arrs) > 0
+    if isinstance(arrs[0], DictObs):
+        return DictObs.concatenate(arrs)
+    else:
+        return np.concatenate(arrs)
+
+
+def stack_maybe_dictobs(arrs: List[ObsVar]) -> ObsVar:
+    """Stacks a list of observations appropriately (depending on type)."""
+    assert len(arrs) > 0
+    if isinstance(arrs[0], DictObs):
+        return DictObs.stack(arrs)
+    else:
+        return np.stack(arrs)
+
+
+# the following overloads have a type error as a DictObs matches both definitions, but
+# the return types are incompatible. Ideally T would exclude DictObs but that's not
+# possible.
+@overload
+def maybe_unwrap_dictobs(  # type: ignore[misc]
+    maybe_dictobs: DictObs,
+) -> Dict[str, np.ndarray]:
+    ...
+
+
+@overload
+def maybe_unwrap_dictobs(maybe_dictobs: T) -> T:
+    ...
+
+
+def maybe_unwrap_dictobs(maybe_dictobs):
+    """Unwraps if a DictObs, otherwise returns the object."""
+    if isinstance(maybe_dictobs, DictObs):
+        return maybe_dictobs.unwrap()
+    else:
+        if not isinstance(maybe_dictobs, (np.ndarray, th.Tensor, int)):
+            warnings.warn(f"trying to unwrap object of type {type(maybe_dictobs)}")
+        return maybe_dictobs
+
+
+@overload
+def maybe_wrap_in_dictobs(obs: Union[Dict[str, np.ndarray], DictObs]) -> DictObs:
+    ...
+
+
+@overload
+def maybe_wrap_in_dictobs(obs: np.ndarray) -> np.ndarray:
+    ...
+
+
+def maybe_wrap_in_dictobs(
+    obs: Union[Dict[str, np.ndarray], np.ndarray, DictObs],
+) -> Observation:
+    """Converts an observation into a DictObs, if necessary."""
+    if isinstance(obs, dict):
+        return DictObs(obs)
+    else:
+        if not isinstance(obs, (np.ndarray, DictObs, float, int)):
+            warnings.warn(f"tried to wrap {type(obs)} as an observation")
+        return obs
+
+
+def map_maybe_dict(fn, maybe_dict):
+    """Applies fn to all values a dictionary, or to the value itself if not a dict."""
+    if isinstance(maybe_dict, dict):
+        return {k: fn(v) for k, v in maybe_dict.items()}
+    else:
+        return fn(maybe_dict)
+
+
+# TODO: maybe should support DictObs?
```

Comment:
Nit: https://google.github.io/styleguide/pyguide.html#312-todo-comments encouraged to leave username or issue in TODO, e.g. `TODO(nix): `, so future developers know where to find more info.